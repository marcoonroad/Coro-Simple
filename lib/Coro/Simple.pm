use v6;

unit module Coro::Simple;

sub coro (&block) is export {
    return sub (*@params) {
        # we explore the lazy evaluation property
        # to evaluate the computation on demand
        my @yields := (gather { block |@params }).list;
        my $index   = 0;
        # generator
        return sub ( ) {
            my $result;
            if defined @yields[ $index++ ] {
                $result = @yields[ $index - 1 ];
            }
            # otherwise, if the coroutine is dead
            else {
        	$index -= 1; # disallows further lookups
                $result = False;
            }
            return $result;
        }
    }
}

sub yield ($value) is export { take $value }

sub suspend( ) is export { take True }

# purpose:
# to check if generated value was returned from yield triggering (rather end of computation)
sub ensure (&block, $value) is export {
    if ($value ~~ Bool) && (!$value) {
        return block;
    }
    return $value;
}

# deprecated function
sub assert (&block, $value) is export {
    warn "[ Deprecated function ]\n" ~
        "This function will be removed in a future version. Please use 'ensure' instead.";
    return ensure &block, $value;
}

sub from (&generator) is export {
    my $temp = generator; # obtains the first value

    return gather {
        while ($temp !~~ Bool) || (?$temp) {
            take $temp;
            $temp = generator;
        }
    };
}

# end of module
