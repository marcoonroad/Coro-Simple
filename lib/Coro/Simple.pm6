use v6;

unit module Coro::Simple;

sub coro (&block) is export {
  my $coroutine-thread = $*THREAD;

  return sub (*@params) {
    # we explore the lazy evaluation property
    # to evaluate the computation on demand
    my @yields := (gather { block |@params }).list;
    my $index   = 0;
    # generator
    return sub ( ) {
      unless $*THREAD.id == $coroutine-thread.id {
        die "Yield called outside the owner coroutine!";
      }

      my $result;
      if defined @yields[ $index++ ] {
        $result = @yields[ $index - 1 ];
      }
      # otherwise, if the coroutine is dead
      else {
	      $index -= 1; # disallows further lookups
        $result = False;
      }
      return $result;
    }
  }
}

sub yield ($value) is export { take $value }

sub suspend( ) is export { take True }

# purpose:
# to check if generated value was returned from yield triggering (rather end of computation)
sub ensure (&block, $value) is export {
  if ($value ~~ Bool) && (!$value) {
    return block;
  }
  return $value;
}

sub from (&generator) is export {
  my $temp = generator; # obtains the first value

  return gather {
    while ($temp !~~ Bool) || (?$temp) {
      take $temp;
      $temp = generator;
    }
  };
}

# end of module
