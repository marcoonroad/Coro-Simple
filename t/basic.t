#!/usr/bin/perl6

use v6;

use Test;
use Coro::Simple;

plan 6;

my $expected = slurp './t/expected/basic.out';
my $handler = open './t/expected/basic.out', :w;

my $coro = coro {
  my $cnt = 1;
  $handler.say: "cnt has: $cnt";
  yield $cnt;

  $cnt += 1;
  $handler.say: "cnt (again) has: $cnt";
  yield [ $cnt, $cnt ];

  $cnt = "Now I'm a string!";
  $handler.say: "Now, cnt is a string? { $cnt ~~ Str }";
  yield $cnt;

  $handler.say: "Hi, folks!";
  $handler.say: "Hello ", "World!" ;
  suspend;

  $handler.say: "See ya later";
  yield [ "Bye-bye!".comb ];
};

my $gen    = $coro( );
my $result = $gen( );

while ($result !~~ Bool) || (?$result) {
  ok((defined $result), "generator result is non-null");
  $result = $gen( );
}

$handler.close;
my $current = slurp './t/expected/basic.out';
is $expected, $current, "generated logs must match";

# end of test
