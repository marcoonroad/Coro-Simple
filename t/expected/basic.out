cnt has: 1
cnt (again) has: 2
Now, cnt is a string? True
Hi, folks!
Hello World!
See ya later
